_tools/types/main.go

```diff
@@ -0,0 +1,341 @@
+		}
+	}
+
+	uastTypes := findUASTTypesInSDK()
+	for _, driver := range drivers {
+		if err := analyzeFixtures(driver); err != nil {
+			return err
+		}
+		analyzeCode(driver, uastTypes)
+	}
+
+	formatMarkdownTable(drivers, uastTypes)
+	return nil
+}
+
+type driverStats struct {
+	url          string
+	language     string
+	path         string
+	skip         bool
+	fixturesNum  int
+	fixturesUast map[string]int
+	codeUast     map[string]int
+}
+
+// listDrivers lists all available drivers.
+func listDrivers() ([]*driverStats, error) {
+	fmt.Fprintf(os.Stderr, "discovering all available drivers\n")
+	langs, err := discovery.OfficialDrivers(context.TODO(), &discovery.Options{
+		NoStatic:      true,
+		NoMaintainers: true,
+		NoSDKVersion:  true,
+		NoBuildInfo:   true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	drivers := make([]*driverStats, 0, len(langs))
+	for _, l := range langs {
+		drivers = append(drivers, &driverStats{
+			language:     l.Language,
+			url:          l.RepositoryURL(),
+			path:         l.RepositoryURL()[strings.LastIndex(l.RepositoryURL(), "/"):],
+			fixturesUast: make(map[string]int),
+			codeUast:     make(map[string]int),
+		})
+	}
+	fmt.Fprintf(os.Stderr, "%d drivers available on-line\n", len(langs))
+	return drivers, nil
+}
+
+// maybeCloneOrPullAll either clones repos to path in local FS or, if already preset,
+// pulls the latest master for each of them.
+func maybeCloneOrPullAll(drivers []*driverStats) error {
+	fmt.Fprintf(os.Stderr, "cloning %d drivers to %s\n", len(drivers), reposRootPath)
+	err := os.MkdirAll(reposRootPath, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	var (
+		pf = make(chan int, 3) // fetches to run in parallel
+		wg sync.WaitGroup
+	)
+	for i := range drivers {
+		wg.Add(1)
+		go func(driver *driverStats) {
+			defer wg.Done()
+			maybeCloneOrPullThrottled(driver, pf)
+		}(drivers[i])
+	}
+	wg.Wait()
+	return nil
+}
+
+// maybeCloneOrPullThrottled is concurent-friendly verison of maybeCloneOrPull.
+// It "throttles" git operarions, based on given chanel capacity.
+func maybeCloneOrPullThrottled(d *driverStats, concurent chan int) error {
+	concurent <- 1
+	defer func() {
+		<-concurent
+	}()
+
+	return maybeCloneOrPull(d)
+}
+
+func maybeCloneOrPull(d *driverStats) error {
+	repoPath := path.Join(reposRootPath, d.path)
+	_, err := os.Stat(repoPath)
+	if os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "%s does not exist, cloning from %s\n", repoPath, d.url)
+		cmd := exec.Command("git", "clone", d.url+".git")
+		cmd.Dir = reposRootPath
+		err = cmd.Run()
+		if err != nil {
+			return err
+		}
+		return nil
+	} else if err != nil {
+		fmt.Fprintf(os.Stderr, "error: %s\n", err)
+		return err
+	}
+
+	fmt.Fprintf(os.Stderr, "%s dir exists, will git pull instead\n", repoPath)
+	cmd := exec.Command("git", "pull", "origin", "master")
+	cmd.Dir = repoPath
+	err = cmd.Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type uastType struct {
```

Comment:
Could this be more simply:

```go
type uastType string

func (u uastType) isUsedIn(pkg string) bool { return … }
```
?

It does not appear to require mutability, and it's fine to use a plain string as the structure.

Reply:
The needed fields will be decided at the time of implementing the `analyzeCode`, will add a comment.